File: java/framework/core/src/main/java/cn/wjybxx/bigcat/fx/DefaultMainModule.java
/*
 *  Copyright 2023-2024 wjybxx
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to iBn writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package cn.wjybxx.bigcat.fx;

import cn.wjybxx.concurrent.IAgentEventHandler;
import cn.wjybxx.concurrent.IEventLoop;
import cn.wjybxx.concurrent.IEventLoopAgent;
import com.google.inject.Inject;
import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;

import java.util.Objects;

/**
 * @author wjybxx
 * date - 2023/12/23
 */
public class DefaultMainModule implements IEventLoopAgent<WorkerEvent> {

    @Inject
    protected TimeModule timeModule;
    /** 帧循环间隔 */
    private int frameInterval = 30;

    /** 主循环前时间戳 - 用于计算帧耗时等 */
    protected long timeBeforeMainLoop;
    /** 主循环后时间戳 */
    protected long timeAfterMainLoop;
    /** 上一次主循环耗时 */
    protected long mainLoopTimeSpan;
    //
    /** 事件循环 */
    protected Worker worker;
    /** 事件循环的事件处理器 */
    protected final Int2ObjectMap<IAgentEventHandler<? super WorkerEvent>> handlerMap = new Int2ObjectOpenHashMap<>(20);

    // region 事件
    @Override
    public void inject(IEventLoop eventLoop, long consumerId) {
        this.worker = (Worker) eventLoop;
    }

    @Override
    public void subscribe(int type, IAgentEventHandler<? super WorkerEvent> handler) {
        Objects.requireNonNull(handler, "handler");
        if (handlerMap.containsKey(type)) {
            throw new IllegalArgumentException("type: " + type);
        }
        handlerMap.put(type, handler);
    }

    @Override
    public void onEvent(long sequence, WorkerEvent event) throws Exception {
        IAgentEventHandler<? super WorkerEvent> handler = handlerMap.get(event.getType());
        if (handler != null) {
            handler.onEvent(sequence, event);
        }
    }
    // endregion

    // region 主循环

    /** 获取帧间隔参数 */
    public int getFrameInterval() {
        return frameInterval;
    }

    /** 设置帧间隔参数 */
    public void setFrameInterval(int frameInterval) {
        if (frameInterval <= 0) throw new IllegalArgumentException("frameInterval: " + frameInterval);
        this.frameInterval = frameInterval;
    }

    /** 获取前一次主循环耗时 -- 或当前主循环结束后查看本次耗时 */
    public long getMainLoopTimeSpan() {
        return mainLoopTimeSpan;
    }

    /** 实时的主循环耗时 */
    public long mainLoopElapsed() {
        return System.currentTimeMillis() - timeBeforeMainLoop;
    }

    @Override
    public void beforeEventLoopStart() {
        timeModule.start(System.currentTimeMillis());
        timeBeforeMainLoop = timeAfterMainLoop = timeModule.getTime();
    }

    @Override
    public boolean checkMainLoop(long threadTime) {
        return System.currentTimeMillis() - timeModule.getTime() >= frameInterval;
    }

    @Override
    public void beforeMainLoop(long threadTime) {
        long timeMillis = System.currentTimeMillis();
        timeModule.update(timeMillis);
        timeBeforeMainLoop = timeMillis;
    }

    @Override
    public void afterMainLoop(long threadTime) {
        timeAfterMainLoop = System.currentTimeMillis();
        mainLoopTimeSpan = timeAfterMainLoop - timeBeforeMainLoop;
    }

    // endregion

}

File: csharp/Wjybxx.BigCat.Core/src/Fx/DefaultMainModule.cs
#region LICENSE

// Copyright 2025 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Collections.Generic;
using Wjybxx.Commons;
using Wjybxx.Commons.Concurrent;
using Wjybxx.Commons.Inject;
using Wjybxx.Commons.Inject.Attributes;

namespace Wjybxx.BigCat.Fx
{
public class DefaultMainModule : EventLoopModule, IEventLoopAgent<WorkerEvent>
{
#nullable disable
    /** 时间模块由主模块驱动 */
    [Inject]
    protected TimeModule timeModule;
    /** 帧循环间隔 */
    private int frameInterval = 30;

    /** 主循环前时间戳 - 用于计算帧耗时等 */
    protected long timeBeforeMainLoop;
    /** 主循环后时间戳 */
    protected long timeAfterMainLoop;
    /** 上一次主循环耗时 */
    protected long mainLoopTimeSpan;
    //
    /** 事件循环 */
    protected Worker worker;
    /** 事件循环的事件处理器 */
    protected readonly Dictionary<int, IAgentEventHandler<WorkerEvent>> handlerMap = new(20);

    #region 事件

    public void Inject(IEventLoop eventLoop, long consumerId) {
        this.worker = (Worker)eventLoop;
    }

    public void Subscribe(int type, IAgentEventHandler<WorkerEvent> handler) {
        if (handler == null) throw new ArgumentNullException(nameof(handler));
        if (handlerMap.ContainsKey(type)) {
            throw new ArgumentException("type: " + type);
        }
        handlerMap[type] = handler;
    }

    public void OnEvent(long sequence, ref WorkerEvent rawEvent) {
        if (handlerMap.TryGetValue(rawEvent.Type, out IAgentEventHandler<WorkerEvent> handler)) {
            handler.OnEvent(sequence, ref rawEvent);
        }
    }

    #endregion

    #region 主循环

    /// <summary>
    /// 帧间隔
    /// </summary>
    public int FrameInterval {
        get => frameInterval;
        set {
            if (value < 1) throw new ArgumentException("frameInterval: " + frameInterval);
            frameInterval = value;
        }
    }

    /// <summary>
    /// 获取前一次主循环耗时 -- 或当前主循环结束后查看本次耗时
    /// </summary>
    /// <value></value>
    public long MainLoopTimeSpan => mainLoopTimeSpan;

    /// <summary>
    /// 实时的主循环耗时
    /// 注意：我们这里需要使用Unix时间戳，而不是系统Tick对应的毫秒数
    /// </summary>
    public long MainLoopElapsed => DatetimeUtil.CurrentEpochMillis() - timeBeforeMainLoop;

    public void BeforeEventLoopStart() {
        timeModule.Start(DatetimeUtil.CurrentEpochMillis());
        timeBeforeMainLoop = timeAfterMainLoop = timeModule.Time;
    }

    public bool CheckMainLoop(long threadTime) {
        return DatetimeUtil.CurrentEpochMillis() - timeModule.Time >= frameInterval;
    }

    public void BeforeMainLoop(long threadTime) {
        long timeMillis = DatetimeUtil.CurrentEpochMillis();
        timeModule.Update(timeMillis);
        timeBeforeMainLoop = timeMillis;
    }

    public void AfterMainLoop(long threadTime) {
        timeAfterMainLoop = DatetimeUtil.CurrentEpochMillis();
        mainLoopTimeSpan = timeAfterMainLoop - timeBeforeMainLoop;
    }

    #endregion
}
}